File: WolfDen/wwwroot/ts/MoveResolver.ts
namespace TurnBasedBoardGame {
    export class MoveResolver {
        
        Game: Game;
        CommandCompletedEvent: any = {};
        MoveInProgress: boolean = false;
        MoveQueue: any[] = [];
        SanityQueue: SanityResult[] = [];
        CurrentSanity: SanityResult;

        constructor() {

        }

        SetMoveInProgress() {
            this.MoveInProgress = true;
        }

        MoveCompleted() {
            if (!SanityCheck(this.Game, this.CurrentSanity)) {
                this.Game.Hub.invoke("sanityCheckFailed");
            }

            this.MoveInProgress = false;
            if (this.MoveQueue.length > 0) {
                let move = this.MoveQueue.shift();
                let sanityResult = this.SanityQueue.shift();
                this.ExecuteMove(move, sanityResult);
            }
        }

        ExecuteMove(move: any, sanityResult: SanityResult): boolean {

            // Check if a move is already in progress
            if (this.MoveInProgress) {
                this.MoveQueue.push(move);
                this.SanityQueue.push(sanityResult);
                return false;
            }

            console.log("Execute move" + sanityResult.MoveNumber + ": " + JSON.stringify(move));
            this.CurrentSanity = sanityResult;
            let succes = false;

            //  Check if move is for a piece
            if (this.Game.Pieces[move.id]) {
                let piece = this.Game.Pieces[move.id] as Piece;
                succes = piece.ExecuteMove(move.data);
            }

            // Check if move is for a tile
            if (this.Game.Board.Nodes[move.id]) {
                let node = this.Game.Board.Nodes[move.id] as IExecuteMove;
                succes = node.ExecuteMove(move.data);
            }

            // Call command completed event if move was succesfull
            if (succes) {
                for (let key in this.CommandCompletedEvent) {
                    if (typeof this.CommandCompletedEvent[key] === "function")
                        this.CommandCompletedEvent[key]();
                }
                if (!this.MoveInProgress)
                    this.MoveCompleted();
            }

            return false;
        }
    }
}

File: WolfDen/Game/MoveResolver.cs
using System;
using Newtonsoft.Json;
using WolfDen.Game.Interfaces;

namespace WolfDen.Game
{
    public class MoveResolver
    {
        [JsonIgnore]
        public dynamic currentMove;
        public Game Game { get; set; }
        public string Class { get { return GetType().Name; } }

        public MoveResolver(Game game)
        {
            Game = game;
        }

        public bool ExecuteMove(string userId, dynamic data)
        {
            try
            {
                string id = data.id;
                Piece piece;

                if (Game.Pieces.TryGetValue(id, out piece) && piece.Owner.Id == userId)
                    return piece.ExecuteMove(data.data);
                if (Game.Board.Nodes.ContainsKey(id))
                    return (Game.Board.Nodes[id] as IExecuteMove).ExecuteMove(data.data);
                return false;
            }
            catch (Exception)
            {
                return false;
            }
        }
    }
}